### PR TITLE
Refactored the HTTP Client Implementation

### PR DESCRIPTION
**Pull request by hiranya911**
This code has become rather long and complex over the months. Therefore I'm refactoring it to make the code more modular and easier to follow:

* Added an `AsyncHttpCall` class to handle a single HTTP interaction.
* The old `sendRequest()` method has been broken down into a bunch of smaller methods, and moved into `AsyncHttpCall`.
* Verified that all the existing unit tests pass, and added 2 new test cases to get 100% test coverage on the HTTP client code.

### CODE REVIEW

File: src/utils/api-request.ts [hunk 1]
```diff
@@ -268,233 +268,353 @@ export function parseHttpResponse(
...
+      if (typeof headers['content-type'] === 'undefined') {
+        headers['content-type'] = 'application/json;charset=utf-8';
+      }
+    } else if (validator.isString(this.data)) {
+      data = Buffer.from(this.data as string, 'utf-8');
+    } else {
+      throw new Error('Request data must be a string, a Buffer or a json serializable object');
+    }
+
+    // Add Content-Length header if data exists
```
Comment (bojeil-google) [inline src/utils/api-request.ts]: Nit: exists.

Comment (hiranya911) [reply src/utils/api-request.ts]: Done

File: src/utils/api-request.ts [hunk 2]
```diff
@@ -268,233 +268,353 @@ export function parseHttpResponse(
...
+    return new AsyncHttpCall(config).promise;
+  }
 
-        respStream.on('error', (err) => {
-          if (req.aborted) {
-            return;
-          }
-          reject(enhanceError(err, config, null, req));
-        });
+  constructor(config: HttpRequestConfig) {
```
[the remark this answers is not in the document]
Comment (hiranya911) [reply src/utils/api-request.ts]: Good call. Done.

**Comment by hiranya911** [comment]
Thanks @bojeil-google. I've made the suggested changes.

**Pull request merged**